### PR TITLE
Dynamically generate controllers based on contents of assets

### PR DESCRIPTION
**Pull request by joelanford**

### CODE REVIEW

File: cmd/cluster-olm-operator/main.go [hunk 1]
```diff
@@ -369,3 +182,104 @@ func replaceImageHook(placeholder string, desiredImageEnvVar string) deploymentc
...
+				Resource:  restMapping.Resource.Resource,
+				Namespace: manifest.GetNamespace(),
+				Name:      manifest.GetName(),
+			})
+
+			if manifest.GetKind() == "Deployment" {
+				controllerName := fmt.Sprintf("%sDeployment%s",
+					namePrefix,
+					strings.ReplaceAll(titler.String(manifest.GetName()), "-", ""),
+				)
```
[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/cluster-olm-operator/main.go:244]: I'm tempted to say we do something that works for 4.14 TP, then figure out beyond how to handle renames and/or drops?

File: cmd/cluster-olm-operator/main.go [hunk 2]
```diff
@@ -369,3 +182,104 @@ func replaceImageHook(placeholder string, desiredImageEnvVar string) deploymentc
...
+				return nil
+			}
+			relatedObjects = append(relatedObjects, configv1.ObjectReference{
+				Group:     restMapping.GroupVersionKind.Group,
+				Resource:  restMapping.Resource.Resource,
+				Namespace: manifest.GetNamespace(),
+				Name:      manifest.GetName(),
+			})
+
+			if manifest.GetKind() == "Deployment" {
```
Comment (ncdc) [inline cmd/cluster-olm-operator/main.go:240]: I'd make sure it's `apps` too

File: cmd/cluster-olm-operator/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (joelanford) [reply cmd/cluster-olm-operator/main.go:244]: Fine with me. I'll create an issue to keep track of this.

File: cmd/cluster-olm-operator/main.go [hunk 3]
```diff
@@ -369,3 +182,105 @@ func replaceImageHook(placeholder string, desiredImageEnvVar string) deploymentc
 		return []byte(newDeployment), nil
 	}
 }
+
+type controllerBuilder struct {
```
Comment (ncdc) [inline cmd/cluster-olm-operator/main.go:186]: Doesn't necessarily need to be in this PR, but I think it's time to split this out into a separate file and possibly package.

File: cmd/cluster-olm-operator/main.go [hunk 4]
```diff
@@ -369,3 +182,105 @@ func replaceImageHook(placeholder string, desiredImageEnvVar string) deploymentc
...
+	assetsFS fs.FS
+	cl       *clients.Clients
```
Comment (ncdc) [inline cmd/cluster-olm-operator/main.go:188]:
```suggestion
	clients       *clients.Clients
```

File: cmd/cluster-olm-operator/main.go [hunk 5]
```diff
@@ -369,3 +182,105 @@ func replaceImageHook(placeholder string, desiredImageEnvVar string) deploymentc
...
+	cc       *controllercmd.ControllerContext
```
Comment (ncdc) [inline cmd/cluster-olm-operator/main.go:189]: I'd rename cc to something like ctrlCtx or controllerCtx or controllerContext

File: cmd/cluster-olm-operator/main.go [hunk 6]
```diff
@@ -369,3 +182,105 @@ func replaceImageHook(placeholder string, desiredImageEnvVar string) deploymentc
...
+			}
+			relatedObjects = append(relatedObjects, configv1.ObjectReference{
+				Group:     restMapping.GroupVersionKind.Group,
+				Resource:  restMapping.Resource.Resource,
+				Namespace: manifest.GetNamespace(),
+				Name:      manifest.GetName(),
+			})
+
+			if manifestGVK.Kind == "Deployment" && manifestGVK.Group == "apps" {
+				controllerName := fmt.Sprintf("%sDeployment%s",
```
Comment (ncdc) [inline cmd/cluster-olm-operator/main.go:242]: This probably should be a separate function with tests

Comment (ncdc) [reply cmd/cluster-olm-operator/main.go:242]: And/or think about how we could write this so it's more testable?

**Comment by ncdc** [comment]
Review comments will be in future PR.
/lgtm

**Pull request merged**